### PR TITLE
[MRG] early decode bytes returned by check_output

### PR DESCRIPTION
**Pull request by ogrisel**
Decode bytes strings into unicode strings as soon as we get the output of a subprocess command.

Let's wait for travis to check that this is working properly.

### CODE REVIEW

**Comment by ogrisel** [comment]
Before merging I would like to add another test.

File: clusterlib/tests/test_scheduler.py [hunk 1]
```diff
@@ -133,9 +133,7 @@ def test_queued_or_running_jobs_slurm():
         out = subprocess.check_output(['squeue', '--noheader', '-o', '%j %i',
                                        '-u', user])
         for line in out.splitlines():
-            if sys.version_info[0] == 3:
-                line = line.decode("utf8")
-            name, id_ = line.split()
+            name, id_ = line.decode("utf8").split()
```
Comment (arjoly) [inline clusterlib/tests/test_scheduler.py]: Do we still the decode here?

File: clusterlib/scheduler.py [hunk 2]
```diff
@@ -101,14 +101,15 @@ def _slurm_queued_or_running_jobs(user=None):
...
+            out = out.decode(encoding)
+            out = [l for l in out.splitlines()]
             return out
     except OSError:
         # squeue is not available
         return []
 
 
-def queued_or_running_jobs(user=None):
+def queued_or_running_jobs(user=None, encoding='utf-8'):
```
Comment (arjoly) [inline clusterlib/scheduler.py]: Can you add a docstring for the new parameter and explain the use case to the user?

**Comment by arjoly** [comment]
Nice addition !

File: clusterlib/tests/test_scheduler.py (hunk 1 above)
Comment (ogrisel) [inline clusterlib/tests/test_scheduler.py]: yes, every time we call `check_output`.

**Comment by ogrisel** [comment]
The new test uses non-ascii UTF-8 encoded chars. If both SLURM and SGE accept it I will document the encoding parameter. If not I will refactor the tests to:

- for ASCII encoding
- check that non-ASCII job name fail with a `ValueError` with a meaningful message.

File: clusterlib/tests/test_scheduler.py (hunk 1 above)
Comment (arjoly) [inline clusterlib/tests/test_scheduler.py]: Thanks !

**Comment by ogrisel** [comment]
@arjoly I have updated the docstest to have them pass.

The `submit` function now always return a utf-8 encoded by string. That might be confusing for some users but it makes it always safe to call `os.system` on that whatever the version of Python.

Alternatevely, I could change that so that the `submit` function always return a unicode string. This would probably feel more intutive in most cases. If the user put non-ascii chars in the `job_name` parameter that would be his/her responsability to encode the output of the `submit` to utf-8 prior to passing it to `os.system` or `subprocess.call`. WDYT?

**Comment by ogrisel** [comment]
BTW, both SGE and SLURM seem to accept UTF-8 encoded, non-ascii characters in the job name.

**Comment by arjoly** [comment]
> The submit function now always return a utf-8 encoded by string. That might be confusing for some users but it makes it always safe to call os.system on that whatever the version of Python.

Why it's confusing?
I like this option. For me, it seems to be the user friendly choice and always gives stable output.
In which extent is the backward compatibility preserved?

(strings have still a part of mystery for me).

**Comment by ogrisel** [comment]
> Why it's confusing?

Because users might expect to do (unicode) string operations (such as calling print on it) on them that don't work well in Python 3 without first decoding. For instance to print a byte string for visualization you have to decode it first.

In the current state of this branch, that return encoded byte string you get the following behavior with Python 3:

```
>>> print(submit(command="my_program", job_name="test_job"))
b'echo \'#!/bin/bash\nmy_program\' | qsub -N "test_job" -l h_rt=24:00:00 -l h_vmem=4000M'
>>> print(submit("my_program", job_name="test_job", backend='sge').decode('utf-8'))
echo '#!/bin/bash
my_program' | qsub -N "test_job" -l h_rt=24:00:00 -l h_vmem=4000M
```

Python 2 does it implicitly for you but it can be problematic if the platform specifc encoding used by Python 2 is not the one that was used to encode the string in the first place.

> For me, it seems to be the user friendly choice and always gives stable output.

If we return unicode string (that is Python 2 `unicode` or Python 3 `str`), they are still stable.

> In which extent is the backward compatibility preserved?

If we make the `submit` method return the unicode string you this will be backward compatible

**Comment by ogrisel** [comment]
I pushed a new commit with that change. If you like the current state. I will add a new comment to add the docstring for the `encoding` parameter of the `queued_or_running_jobs`.

**Comment by ogrisel** [comment]
Alright I added the doc. I also updated the users guide to give some insight on how to properly encode the command generated by `submit` when the `job_name` parameters contains non-ASCII chars. I can squash the commits if you wish.

**Comment by arjoly** [comment]
Before a final review, do you also prefer this solution over the unicode one?

**Comment by ogrisel** [comment]
> Before a final review, do you also prefer this solution over the unicode one?

I prefer the current state of this PR.

**Comment by arjoly** [comment]
Can you remove file.txt?

**Comment by arjoly** [comment]
Can you highlight your contribution in the what's new?

**Comment by arjoly** [comment]
Beside the above comment, LGTM!

**Comment by ogrisel** [comment]
Comments addressed. This should be good to merge. Sorry for the leftover `file.txt` file.

**Comment by arjoly** [comment]
Merge by rebase !!!

Thanks @ogrisel for improving clusterlib!!!

**Pull request merged**